### ISSUE TITLE
Optimize / paralllellize building of idris packages

### ISSUE THREAD

**Issue by luc-tielen** [ejconlon/idream#5]
As dicussed in #3 while implementing the various command line actions for idream, there are multiple ways to compile the packages.
(Turned this into a separate issue since other issue thread would get too long..)

A "complex" dependency graph could look as follows:
![example_dependency_graph](https://user-images.githubusercontent.com/3411018/36342361-cf50ef86-13fd-11e8-846e-dde6a0138657.png)

Main 3 options for compiling this hierarchy of packages are:
1. sequentially building all packages after toplogically sorting them
2. use MonadParIO from the [monad-par](https://hackage.haskell.org/package/monad-par-0.3.4.8/docs/Control-Monad-Par.html) package 
3. use actors on top of [async](https://hackage.haskell.org/package/async)

The idea would be to start with option 1 and then later optimize to use option 2 or 3.

Option 1 would be very straight forward by topologically sorting the packages and then compiling each package one by one sequentially.
Option 2 should be a small step upwards from option 1, might be necessary to create some kind of BuildPlan data type that can then be used for expressing the dataflow parallellism better (see also the monad-par docs mentioned above on how to express this kind of parallellism) and subsequently interpreting this data type.
Option 3 would consist of writing a small layer on top of the async package in order to have a simple actor framework. As soon as this would be done, an actor could be created for each package in the hierarchy, ready to listen for incoming messages and act upon them. It could also be possible to use existing actor libs in Haskell but they either offer too many features or not enough (+ they also form extra dependencies for idream itself)

Option 3 has the benefit of being more optimal than option 2 in some cases (for example if E takes so long that even A has to start waiting on B) but it is less deterministic in the order of which it builds things (actors would start building as soon as all dependencies have finished building).
Chosen option depends on what we want to go for, but for now I tend to like the slightly more deterministic approach more.

Concerns raised during discussion were that it might not be possible to invoke idris compiler in parallel on a machine. However the dependency graph / build plan approach should make it safe to built independent packages at same time. If there is indeed a problem inside the compiler itself, an issue should be raised on the idris github page so this gets fixed.
